molecule/provisioners/libvirtprovisioner.py

```diff
@@ -0,0 +1,923 @@
+                    #  i.e. after creating the domain's volume but before booting the domain.
+                    # if any of this instance's interfaces has a key 'address'? or always?
+                    try:
+                        guest = guestfs.GuestFS(python_return_dict=True)
+                        self._manipulate_guest_image(guest, instance)
+                        guest.umount_all()
+                        guest.shutdown()
+                        guest.close()
+                    except Exception, e:
+                        LOG.warning("\nFAILED to manipulate the guest image so could not configure network interfaces: {}".format(e))
+                utilities.print_success("\tCreated instance {}.\n".format(
+                    instance['name']))
+                utilities.print_info("\t{}: booting".format(instance['name']))
+                try:
+                    dom.create()
+                    if wait:
+                        time.sleep(self._boot_wait)
+                        wait = False
+                except libvirt.libvirtError as e:
+                    LOG.error("\nFailed to create/boot {}: {}".format(instance[
+                        'name'], e))
+                    dom.undefine()
+
+    def _manipulate_guest_image(self, guest, instance):
+        """
+        Ensure that each network interface that libvirt creates has
+        (necessarily distro-specific) network configuration that either
+        a) specifies IP config (according to the instance's interfaces' address)
+        OR
+        b) ensures that the guest is configured to DHCP for each interface
+
+        :param: a GuestFS (for this instance/volume)
+        :param: an instance (so we can determine the desired network config)
+        """
+        drive = self._pool_path + '/' + instance['name'] + '.img'
+        guest.add_drive_opts(drive, readonly=0)
+        LOG.info("\tLaunching libguestfs to edit {}".format(drive))
+        guest.launch()
+        # Ask libguestfs to inspect for operating systems.
+        roots = guest.inspect_os()
+        if len(roots) == 0:
+            raise(Error("inspect_vm: no operating systems found"))
+        distro = ''
+        for root in roots:
+            distro = guest.inspect_get_distro(root)
+            # Sort keys by length, shortest first, so that we end up
+            # mounting the filesystems in the correct order.
+            mps = guest.inspect_get_mountpoints(root)
+            def compare(a, b):
+                return len(a) - len(b)
+            for device in sorted(mps.keys(), compare):
+                try:
+                    guest.mount(mps[device], device)
+                except RuntimeError as msg:
+                    print "%s (ignored)" % msg
+        LOG.info("\tlibguestfs found {}".format(distro))
+
+        guest.write("/etc/resolv.conf", "nameserver 8.8.8.8")
+        if distro.startswith('redhat-based') or distro.startswith('rhel'):
```

Comment:
I think this should be completely abstracted away (RHEL/Ubuntu difference) in a completely separate class.
```
config = NetworkConfig(distro, instance['interfaces'])
for interface_path, interface_config in config.get_interfaces_config():
    guest.write(interface_path, interface_config)
```